### PR TITLE
Detailed thermostat setpoints, try 2

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

Adds arguments to `hpxml.rb` to support multiple setback/setup periods:
* `weekday_heating_setpoints`
* `weekend_heating_setpoints`
* `weekday_cooling_setpoints`
* `weekend_cooling_setpoints`

Adds test xml that uses these new arguments (and also has a few hours where cooling setpoint < heating setpoint):
```
<HVACControl>
  <SystemIdentifier id='HVACControl'/>
  <ControlType>programmable thermostat</ControlType>
  <extension>
    <WeekdaySetpointTempsHeatingSeason>64, 64, 64, 64, 64, 64, 64, 74, 74, 66, 66, 66, 66, 66, 66, 66, 66, 68, 68, 68, 68, 68, 64, 64</WeekdaySetpointTempsHeatingSeason>
    <WeekendSetpointTempsHeatingSeason>64, 64, 64, 64, 64, 64, 64, 74, 74, 66, 66, 66, 66, 66, 66, 66, 66, 68, 68, 68, 68, 68, 64, 64</WeekendSetpointTempsHeatingSeason>
    <WeekdaySetpointTempsCoolingSeason>82, 82, 82, 82, 82, 82, 82, 72, 72, 80, 80, 80, 80, 80, 80, 80, 80, 78, 78, 78, 78, 78, 82, 82</WeekdaySetpointTempsCoolingSeason>
    <WeekendSetpointTempsCoolingSeason>82, 82, 82, 82, 82, 82, 82, 72, 72, 80, 80, 80, 80, 80, 80, 80, 80, 78, 78, 78, 78, 78, 82, 82</WeekendSetpointTempsCoolingSeason>
  </extension>
</HVACControl>
```

## Checklist

Not all may apply:

- [x] EPvalidator.xml has been updated
- [x] Tests (and test files) have been updated
- [x] Documentation has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected regression test changes on CI

### CODE REVIEW

**Review by shorowit — COMMENTED** [review]
I suggest we use flat elements instead of nested elements for simplicity, and choose names more consistent with names already in HPXML.

So instead of this:
```
<extension>
  <HeatingSetpointSchedules>
    <Weekday>64, 64, 64, 64, 64, 64, 64, 74, 74, 66, 66, 66, 66, 66, 66, 66, 66, 68, 68, 68, 68, 68, 64, 64</Weekday>
    <Weekend>64, 64, 64, 64, 64, 64, 64, 74, 74, 66, 66, 66, 66, 66, 66, 66, 66, 68, 68, 68, 68, 68, 64, 64</Weekend>
  </HeatingSetpointSchedules>
  <CoolingSetpointSchedules>
    <Weekday>82, 82, 82, 82, 82, 82, 82, 72, 72, 80, 80, 80, 80, 80, 80, 80, 80, 78, 78, 78, 78, 78, 82, 82</Weekday>
    <Weekend>82, 82, 82, 82, 82, 82, 82, 72, 72, 80, 80, 80, 80, 80, 80, 80, 80, 78, 78, 78, 78, 78, 82, 82</Weekend>
  </CoolingSetpointSchedules>
</extension>
```

I suggest this:
```
<extension>
  <WeekdaySetpointTempsHeatingSeason>64, 64, 64, 64, 64, 64, 64, 74, 74, 66, 66, 66, 66, 66, 66, 66, 66, 68, 68, 68, 68, 68, 64, 64</WeekdaySetpointTempsHeatingSeason>
  <WeekendSetpointTempsHeatingSeason>64, 64, 64, 64, 64, 64, 64, 74, 74, 66, 66, 66, 66, 66, 66, 66, 66, 68, 68, 68, 68, 68, 64, 64</WeekendSetpointTempsHeatingSeason>
  <WeekdaySetpointTempsCoolingSeason>82, 82, 82, 82, 82, 82, 82, 72, 72, 80, 80, 80, 80, 80, 80, 80, 80, 78, 78, 78, 78, 78, 82, 82</WeekdaySetpointTempsCoolingSeason>
  <WeekendSetpointTempsCoolingSeason>82, 82, 82, 82, 82, 82, 82, 72, 72, 80, 80, 80, 80, 80, 80, 80, 80, 78, 78, 78, 78, 78, 82, 82</WeekendSetpointTempsCoolingSeason>
</extension>
```

File: HPXMLtoOpenStudio/resources/EPvalidator.xml [hunk 1]
```diff
@@ -672,10 +672,16 @@
...
-      <sch:assert role='ERROR' test='count(h:SetpointTempCoolingSeason) = 1'>Expected 1 element(s) for xpath: SetpointTempCoolingSeason</sch:assert>
+      <sch:assert role='ERROR' test='count(h:SetpointTempHeatingSeason) + count(h:extension/h:WeekdaySetpointTempsHeatingSeason) = 1'>Expected 1 element(s) for xpath: SetpointTempHeatingSeason | WeekdaySetpointTempsHeatingSeason</sch:assert>
+      <sch:assert role='ERROR' test='count(h:SetpointTempHeatingSeason) + count(h:extension/h:WeekendSetpointTempsHeatingSeason) = 1'>Expected 1 element(s) for xpath: SetpointTempHeatingSeason | WeekendSetpointTempsHeatingSeason</sch:assert>
+      <sch:assert role='ERROR' test='count(h:extension/h:WeekdaySetpointTempsHeatingSeason) + count(h:extension/h:WeekendSetpointTempsHeatingSeason) = 0 or count(h:extension/h:WeekdaySetpointTempsHeatingSeason) + count(h:extension/h:WeekendSetpointTempsHeatingSeason) = 2'>Expected 0 or 2 element(s) for xpath: WeekdaySetpointTempsHeatingSeason | WeekendSetpointTempsHeatingSeason</sch:assert>
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/EPvalidator.xml:677]: This line doesn't seem needed. If you pass the two tests above, you will by definition pass this test, no?

File: HPXMLtoOpenStudio/resources/EPvalidator.xml [hunk 2]
```diff
@@ -672,10 +672,16 @@
...
+      <sch:assert role='ERROR' test='count(h:SetbackTempHeatingSeason) + count(h:extension/h:WeekdaySetpointTempsHeatingSeason) &lt;= 1'>Expected 0 or 1 element(s) for xpath: SetbackTempHeatingSeason | WeekdaySetpointTempsHeatingSeason</sch:assert> <!-- See [HVACControlType=HeatingSetback] -->
+      <sch:assert role='ERROR' test='count(h:SetbackTempHeatingSeason) + count(h:extension/h:WeekendSetpointTempsHeatingSeason) &lt;= 1'>Expected 0 or 1 element(s) for xpath: SetbackTempHeatingSeason | WeekdaySetpointTempsHeatingSeason</sch:assert> <!-- See [HVACControlType=HeatingSetback] -->
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/EPvalidator.xml:679]: Weekday -> Weekend

File: HPXMLtoOpenStudio/resources/EPvalidator.xml [hunk 3]
```diff
@@ -672,10 +672,16 @@
...
+      <sch:assert role='ERROR' test='count(h:SetpointTempCoolingSeason) + count(h:extension/h:WeekdaySetpointTempsCoolingSeason) = 1'>Expected 1 element(s) for xpath: SetpointTempCoolingSeason | WeekdaySetpointTempsCoolingSeason</sch:assert>
+      <sch:assert role='ERROR' test='count(h:SetpointTempCoolingSeason) + count(h:extension/h:WeekendSetpointTempsCoolingSeason) = 1'>Expected 1 element(s) for xpath: SetpointTempCoolingSeason | WeekendSetpointTempsCoolingSeason</sch:assert>
+      <sch:assert role='ERROR' test='count(h:extension/h:WeekdaySetpointTempsCoolingSeason) + count(h:extension/h:WeekendSetpointTempsCoolingSeason) = 0 or count(h:extension/h:WeekdaySetpointTempsCoolingSeason) + count(h:extension/h:WeekendSetpointTempsCoolingSeason) = 2'>Expected 0 or 2 element(s) for xpath: WeekdaySetpointTempsCoolingSeason | WeekendSetpointTempsCoolingSeason</sch:assert>
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/EPvalidator.xml:682]: Same comment as above, shouldn't be needed.

File: HPXMLtoOpenStudio/resources/hpxml.rb [hunk 4]
```diff
@@ -2941,6 +2943,35 @@ def delete
 
     def check_for_errors
       errors = []
+
+      if not @weekday_heating_setpoints.nil?
+        vals = @weekday_heating_setpoints.split(', ').map { |i| i.to_f }
+        if vals.length != 24
+          errors << "Weekday Heating Setpoints (#{vals.length}) must be of length 24."
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml.rb:2950]: Can we be more explicit with these error messages? How about "... must be a comma-separated list of 24 numbers."?

File: HPXMLtoOpenStudio/resources/hpxml.rb [hunk 5]
```diff
@@ -2958,9 +2989,19 @@ def to_oga(doc)
...
-      XMLHelper.add_extension(hvac_control, 'SetupStartHourCooling', to_integer(@cooling_setup_start_hour)) unless @cooling_setup_start_hour.nil?
-      XMLHelper.add_extension(hvac_control, 'CeilingFanSetpointTempCoolingSeasonOffset', to_float(@ceiling_fan_cooling_setpoint_temp_offset)) unless @ceiling_fan_cooling_setpoint_temp_offset.nil?
+      if (not @heating_setback_start_hour.nil?) || (not @cooling_setup_start_hour.nil?) || (not @ceiling_fan_cooling_setpoint_temp_offset.nil?)
+        extension = XMLHelper.create_elements_as_needed(hvac_control, ['extension'])
+        XMLHelper.add_element(extension, 'SetbackStartHourHeating', to_integer(@heating_setback_start_hour)) unless @heating_setback_start_hour.nil?
+        XMLHelper.add_element(extension, 'SetupStartHourCooling', to_integer(@cooling_setup_start_hour)) unless @cooling_setup_start_hour.nil?
+        XMLHelper.add_element(extension, 'CeilingFanSetpointTempCoolingSeasonOffset', to_float(@ceiling_fan_cooling_setpoint_temp_offset)) unless @ceiling_fan_cooling_setpoint_temp_offset.nil?
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml.rb:2996]: I'm totally confused. Why did you move these to an extension element?

File: HPXMLtoOpenStudio/resources/hpxml.rb [hunk 6]
```diff
@@ -2958,9 +2989,19 @@ def to_oga(doc)
...
+      end
+      if (not @weekday_heating_setpoints.nil?) || (not @weekend_heating_setpoints.nil?) || (not @weekday_cooling_setpoints.nil?) || (not @weekday_cooling_setpoints.nil?)
+        extension = XMLHelper.create_elements_as_needed(hvac_control, ['extension'])
+        XMLHelper.add_element(extension, 'WeekdaySetpointTempsHeatingSeason', @weekday_heating_setpoints) unless @weekday_heating_setpoints.nil?
+        XMLHelper.add_element(extension, 'WeekendSetpointTempsHeatingSeason', @weekend_heating_setpoints) unless @weekend_heating_setpoints.nil?
+        XMLHelper.add_element(extension, 'WeekdaySetpointTempsCoolingSeason', @weekday_cooling_setpoints) unless @weekday_cooling_setpoints.nil?
+        XMLHelper.add_element(extension, 'WeekendSetpointTempsCoolingSeason', @weekend_cooling_setpoints) unless @weekday_cooling_setpoints.nil?
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hpxml.rb:3003]: weekday -> weekend

**Review by shorowit — COMMENTED** [review]
Some initial comments.

File: HPXMLtoOpenStudio/resources/hpxml.rb (hunk 5 above)
Comment (joseph-robertson) [reply HPXMLtoOpenStudio/resources/hpxml.rb:2996]: I didn't: https://github.com/NREL/OpenStudio-HPXML/blob/master/HPXMLtoOpenStudio/resources/hpxml.rb#L2961-L2963.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 7]
```diff
@@ -1375,6 +1385,29 @@ def self.apply_setpoints(model, runner, weather, hvac_control, living_zone)
...
+    if (not hvac_control.weekday_cooling_setpoints.nil?) && (not hvac_control.weekend_cooling_setpoints.nil?)
+      clg_weekday_setpoints = hvac_control.weekday_cooling_setpoints.split(', ').map { |i| i.to_f }
+      clg_weekday_setpoints = [clg_weekday_setpoints] * 12
+      if not clg_ceiling_fan_offset.nil?
+        HVAC.get_default_ceiling_fan_months(weather).each_with_index do |operation, m|
+          next unless operation == 1
+
+          clg_weekday_setpoints[m] = [clg_weekday_setpoints[m], Array.new(24, clg_ceiling_fan_offset)].transpose.map { |i| i.reduce(:+) }
+        end
+      end
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hvac.rb:1398]: Duplicated code (and twice at that! here and below). 👎 

Please refactor the code so that the ceiling fan offset logic only needs to be applied once.

File: tasks.rb [hunk 8]
```diff
@@ -287,6 +287,7 @@ def create_hpxmls
     'base-hvac-none.xml' => 'base.xml',
     'base-hvac-portable-heater-gas-only.xml' => 'base.xml',
     'base-hvac-programmable-thermostat.xml' => 'base.xml',
+    'base-hvac-programmable-thermostat-multiple-periods.xml' => 'base.xml',
```
Comment (shorowit) [inline tasks.rb:290]: I suggest just calling it 'base-hvac-programmable-thermostat-detailed.xml'

**Review by shorowit — COMMENTED** [review]
In general, the approach looks good. Just needs a little cleanup.

File: HPXMLtoOpenStudio/resources/hpxml.rb [hunk 9]
```diff
@@ -2944,6 +2946,35 @@ def delete
 
     def check_for_errors
       errors = []
+
+      if not @weekday_heating_setpoints.nil?
+        vals = @weekday_heating_setpoints.split(', ').map { |i| i.to_f }
```
[the remark this answers is not in the document]
Comment (joseph-robertson) [reply HPXMLtoOpenStudio/resources/hpxml.rb:2951]: Good point, you are right; we don't need this code. When I remove this code, and just try running `hvac-base-programmable-thermostat-detailed.xml` with `WeekdaySetpointTempsHeatingSeason` set to length of 23 instead of 24, I get in the run.log: `Error: A 12-element array with 24-element arrays of numbers must be entered for the schedule.`

File: HPXMLtoOpenStudio/resources/hvac.rb (hunk 7 above)
Comment (joseph-robertson) [reply HPXMLtoOpenStudio/resources/hvac.rb:1398]: Good catch.

File: HPXMLtoOpenStudio/resources/hpxml.rb (hunk 5 above)
Comment (joseph-robertson) [reply HPXMLtoOpenStudio/resources/hpxml.rb:2999]: This might be an artifact of when I had separate extension elements for Heating vs Cooling fields.

**Review by shorowit — APPROVED** [review]
Looks good to go. Will double-check final CI results before merging.

File: docs/source/hpxml_to_openstudio.rst [hunk 10]
```diff
@@ -627,28 +627,33 @@ HPXML HVAC Control
...
-If there is a heating setback, it is defined with:
+Setpoints can be provided using either simple inputs or detailed inputs:
 
-- ``SetbackTempHeatingSeason``: Temperature during heating setback
-- ``extension/SetbackStartHourHeating``: The start hour of the heating setback where 0=midnight and 12=noon
-- ``TotalSetbackHoursperWeekHeating``: The number of hours of heating setback per week
+1. **Simple inputs**
 
-If there is a cooling setup, it is defined with:
+  Provid the heating setpoint (``SetpointTempHeatingSeason``) and cooling setpoint (``SetpointTempCoolingSeason``) elements.
```
Comment (joseph-robertson) [inline docs/source/hpxml_to_openstudio.rst:635]: Provide

**Comment by shorowit** [comment]
Thanks @joseph-robertson
